kernel/src/fs/epoll/entry.rs

```diff
@@ -0,0 +1,381 @@
+};
+
+/// An epoll entry that is contained in an epoll file.
+///
+/// Each epoll entry can be added, modified, or deleted by the `EpollCtl` command.
+pub(super) struct Entry {
+    // The file descriptor and the file.
+    key: EntryKey,
+    // The event masks and flags.
+    inner: Mutex<Inner>,
+    // The observer that receives events.
+    //
+    // Keep this in a separate `Arc` to avoid dropping `Entry` in the observer callback, which may
+    // cause deadlocks.
+    observer: Arc<Observer>,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+pub(super) struct EntryKey {
+    fd: FileDesc,
+    file: KeyableWeak<dyn FileLike>,
+}
+
+impl From<(FileDesc, KeyableWeak<dyn FileLike>)> for EntryKey {
+    fn from(value: (FileDesc, KeyableWeak<dyn FileLike>)) -> Self {
+        Self {
+            fd: value.0,
+            file: value.1,
+        }
+    }
+}
+
+impl From<(FileDesc, &Arc<dyn FileLike>)> for EntryKey {
+    fn from(value: (FileDesc, &Arc<dyn FileLike>)) -> Self {
+        Self {
+            fd: value.0,
+            file: KeyableWeak::from(Arc::downgrade(value.1)),
+        }
+    }
+}
+
+struct Inner {
+    event: EpollEvent,
+    flags: EpollFlags,
+    poller: PollHandle,
+}
+
+impl Entry {
+    /// Creates a new epoll entry associated with the given epoll file.
+    pub(super) fn new(
+        fd: FileDesc,
+        file: KeyableWeak<dyn FileLike>,
+        ready_set: Arc<ReadySet>,
+    ) -> Arc<Self> {
+        Arc::new_cyclic(|me| {
+            let observer = Arc::new(Observer::new(ready_set, me.clone()));
+
+            let inner = Inner {
+                event: EpollEvent {
+                    events: IoEvents::empty(),
+                    user_data: 0,
+                },
+                flags: EpollFlags::empty(),
+                poller: PollHandle::new(Arc::downgrade(&observer) as _),
+            };
+
+            Self {
+                key: EntryKey { fd, file },
+                inner: Mutex::new(inner),
+                observer,
+            }
+        })
+    }
+
+    /// Gets the file associated with this epoll entry.
+    ///
+    /// Since an epoll entry only holds a weak reference to the file,
+    /// it is possible (albeit unlikely) that the file has been dropped.
+    fn file(&self) -> Option<Arc<dyn FileLike>> {
+        self.key.file.upgrade().map(KeyableArc::into)
+    }
+
+    /// Polls the events of the file associated with this epoll entry.
+    ///
+    /// This method returns `None` if the file is dead. Otherwise, it returns the epoll event (if
+    /// any) and a boolean value indicating whether the entry should be kept in the ready list
+    /// (`true`) or removed from the ready list (`false`).
+    pub(super) fn poll(&self) -> Option<(Option<EpollEvent>, bool)> {
+        let file = self.file()?;
+        let inner = self.inner.lock();
+
+        // There are no events if the entry is disabled.
+        if !self.observer.is_enabled() {
+            return Some((None, false));
+        }
```

Comment:
I think doing the check before holding the lock is correct, but I do not intend to do this.

This is because if we do this, we need to be very careful about the race condition, as mentioned in the method's doc comments below:
```rust
    /// *Caution:* If this method is called without holding the lock of the event masks and flags,
    /// the user must ensure that the behavior is desired with respect to the way the event masks
    /// and flags might be modified concurrently.
    fn is_enabled(&self) -> bool
```

We have to do this in `ReadySet::push` because we cannot take the lock there, so I write detailed comments explaining why race conditions do not lead to bad states:
```rust
        // Note that we cannot take the `Inner` lock because we may be in the callback of the event
        // observer. Doing so will cause dead locks due to inconsistent locking orders.
        //
        // We don't need to take the lock because
        // - `set_enabled` is guaranteed to be visible after its following `FileLike::poll`. This
        //   synchronization relies on the lock in `Subject`. As a result, no events are lost.
        // - Catching spurious events here is always fine because we always check them later before
        //   returning events to the user (in `Entry::poll`).
        if !observer.is_enabled() {
            return;
        }
```

If we do the check in `Entry::poll` in a lockless way, we'll have to argue like this:
```rust
        // There are no events if the entry is disabled.
        //
        // We do not hold the `Inner` lock yet. This check is fine to be placed here because:
        // - `set_enabled` is guaranteed to be visible after its subsequent `ReadySet::push`. This
        //   synchronization is based on the ready list lock. No events are lost.
        // - `reset_enabled` in `Entry::poll` is guaranteed to be visible since we hold an
        //   exclusive `PopGuard`.
        // - `reset_enabled` in `Entry::shutdown` may not be visible. But if it is not, it will be
        //   a race condition in the user program, so we can still return the events.
        if !self.observer.is_enabled() {
            return Some((None, false));
        }
```

While I still think it's right, I'm not sure the benefit outweighs the complexity.

Reply:
Since the change will brings much complexity for discussion about race condition, it's acceptable for me to keep it's current design.